File: CSharp/206-ReverseLinkedList/206ReverseLinkedList.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
///////////////////////////////////////////////////////////////////////////
 // ITERATIVE
 // Time Complexity: O(n) (Analyze Complexity confirmed)
 //// We go through the list once.

 // Space Complexity: O(1) 
 //// We use a constant amoutn of extra space for 3 pointers regardless of the size of the linked list. 
public class Solution {
    public ListNode ReverseList(ListNode head) {
        ListNode prev = null;
        ListNode current = head;
        ListNode nextNode = null;
        while (current != null){
            nextNode = current.next; // save the next node 
            current.next = prev; // reverse the link
            prev = current; 
            current = nextNode;
        }
        return prev; // new head of the reversed list
    }
}

File: Javascript/206-ReverseLinkedList/206ReverseLinkedList.js
// There are two approaches for this. 
// Reverse the string
// Two pointers. 
// Obviously two pointers is longer but it's probably given as a solution because it's an easy introduction to two pointers.
// so I should come up with a javascript implementation for that to make sure I have it


/*
Time Complexity: O(n)
We go through the linked list in a single pass

Space Complexity: O(1)
no matter the size of the linked list we just need the 3 variables prev curr and next.

*/

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function(head) {
    let prev = null;
    let curr = head;
    while (curr !== null){
        let next = curr.next;
        curr.next = prev;
        prev = curr;
        curr = next;
    }
    return prev;
};